### ISSUE TITLE
Draper::Decoratable::ClassMethods#=== can cause unexpected behaviour for non-decorated variables that respond to #object

### ISSUE THREAD

**Issue by DMA57361** [drapergem/draper#705]
#### TLDR

`Draper::Decoratable::ClassMethods#===` will call `super(other.object)` on anything that responds to `object`, not just decorated variables.

Here's a contrived example performed on a rails console that demonstrates some unexpected behaviour. This assumes `User` is a decoratable ActiveRecord class or similar.

```
[1] pry(main)> class NotAUser
[1] pry(main)*   def object
[1] pry(main)*     User.first
[1] pry(main)*   end
[1] pry(main)* end
=> :object
[2] pry(main)> User === NotAUser.new
=> true
```

Hopefully you agree that the result at the end there should be `false`, and the `NotAUser#object` method should never have been called.

---
#### My specific problem

I've been trying to add the prawn gem to one of our applications (it's used to generate PDFs, although that's not really relevant here) and have spend a bit of time fighting with a SystemStackError, which I think is the responsibility of the draper gem.

We patch the class `Time` with `include Draper::Decoratable` so that we can do some extra magic with it in views. At some point, while building the main object of the prawn gem (simply calling `Prawn::Document.new`), one of the internal objects of the prawn gem is used in a case statement where one of the branches checks against `Time`. `Time`, being decoratable in out application, uses the definition of `===` in `Draper::Decoratable::ClassMethods`. Unfortunately, the variable that is the subject of the case statement has an `object` method, and calling this method eventually leads back to the same case statement - so we get a indefinite loop leading to a `SystemStackError`. As with the contrived example at the start, I don't think `.===` should be calling `#object` at all in this case.

Here's a clipping of that backtrace in case it's useful somehow. Note that I'm seeing this in draper v1.3.1 (apparently we're out of date), but have updated to v2.1.0 quickly to see if the problem still exists, and it seems it does. This is Ruby 2.2 and Rails 4.2.

```
  ...
  activesupport (4.2.4) lib/active_support/core_ext/time/calculations.rb:15:in `==='
  pdf-core (0.6.0) lib/pdf/core/pdf_object.rb:63:in `PdfObject'
  pdf-core (0.6.0) lib/pdf/core/pdf_object.rb:88:in `block in PdfObject'
  pdf-core (0.6.0) lib/pdf/core/pdf_object.rb:82:in `each'
  pdf-core (0.6.0) lib/pdf/core/pdf_object.rb:82:in `PdfObject'
  pdf-core (0.6.0) lib/pdf/core/reference.rb:28:in `object'
  draper (2.1.0) lib/draper/decoratable.rb:90:in `==='
  activesupport (4.2.4) lib/active_support/core_ext/time/calculations.rb:15:in `==='
  ...
```

---
#### Possible solution?

By using a monkey patch to redefine `Draper::Decoratable::ClassMethods#===` so that the `other.respond_to?(:object)` check is changed for `other.is_a?(Draper::Decorator)`, my specific problem and the contrived example are fixed. It doesn't _seem_ to have broken anything else in our application, so far...

If there's no unintended side effects (I don't know the inner workings of the gem, so have no idea!) could this, or some safely tested equivalent, be used to more accurately determine when calling `#object` makes sense?

**Comment by codebycliff** [comment]
Thanks for reporting this. This was fixed in https://github.com/drapergem/draper/pull/720.
